### ISSUE TITLE
FR: Switch to react and react native?

### ISSUE THREAD

**Issue by raix** [meteor/meteor#3728]
"Embrace the ecosystem" http://docs.meteor.com/#/basic/sevenprinciples

I'm really happy with Blaze but it seems React is doing the same thing, why not cooperate?
* It got an official component model
* Its fast
* It renders server-side
* It does work natively on ios and android - natively...
* Meteor core team has limited ressources

Native is one killer feature on the way: react native allowing us to write js and use the native ui components.

I've been working on several meteor mobile projects, and just recently bumped to use the official cordova implementation. Its ok, but still cordova with the inconsistencies that follows.

React native is doing something a lot of us have thought about: letting us write js and leverage the native ui. Its actually allowing faster development than regular native development (eg. not having to recompile all the time)

I know some would argue that its only for mobile, but its actually also for desktop, the Meteor build tool supports multiple platforms, this is cool!

We could wrap ios/android native code in our packages instead of trying to fix the cordova plugins all the time...

https://www.youtube.com/watch?v=7rDsRXj9-cU

Just my honest opinion, I think react fits Meteor very well, maybe have a Meteor React distribution for starters to check out popularity...

Kind regards Morten

**Comment by jhartma** [comment]
interesting thought, and definitely something I would try!

**Comment by rclai** [comment]
There is a guy that created this [React.js integration package](https://atmospherejs.com/hipertracker/reactjs). I haven't tried it but it looks like you can actually write JSX and it hooks into the build process, which is nice.

**Comment by raix** [comment]
Theres another guy that build https://github.com/reactjs/react-meteor - I think he is on the core team now?

**Comment by rclai** [comment]
Oh wow, I guess he is. The guy that made the package I mentioned is built using the same code as Ben's.

**Comment by stubailo** [comment]
For a start, I think it would be great to have a well-maintained and useful community react plugin. Is the one you linked to good enough for people to use React as a replacement for Blaze? If not, what is it missing?

Thanks for the feature request! You might be interested in reading our [feature request guidelines](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#feature-requests).

**Comment by rclai** [comment]
Can you ask @benjamn (he's in MDG right?) if [this implementation](https://atmospherejs.com/hipertracker/reactjs) is still good?

**Comment by raix** [comment]
@stubailo I've skimmed the guidelines a while back and just now - not sure if I'm missing something? - This FR is based on the:

> we're not able to work on every single subproject every month

+the reference to the seven principles "Embrace the ecosystem"

+I've contributed to the hackpad Blaze2 regarding this too

I'm personally not seeking to maintain yet a package, the deep integration I would have wished for requires more than a package - making the build tool extendable is another FR.

But lets see when React native is published if its worth migrating to or if its just another Famo.us project claiming to tackle mobile.

Kind regards Morten

**Comment by rclai** [comment]
@raix he's not saying that because "you didn't understand something" he's saying that because MDG requires all employees to state that in every FR post.

**Comment by raix** [comment]
ok, so I just replied on a copy&paste / bot text - I'm too polite to ignore it, @stubailo could you guys consider:
* have a bot comment depending on the labels you set
* have a bot that sets label "feature request" if we put "[FR]" in subject (this way we can help you set the label)
* or  tailing robot messages like instead:

-

*Thanks for the feature request! You might be interested in reading our feature request guidelines.*

**Comment by glasser** [comment]
Well, it's not a bot, it's copied and pasted by a human :)  See https://meteor.hackpad.com/Responding-to-GitHub-Issues-SKE2u3tkSiH for how we try to deal with the issue queue.

The goal of the text is to get across the message that we appreciate the feature request and are happy to track it in our issue tracker, but that there's a giant gap between "Yes, this would be cool" and "Great, we are going to allocate limited labor resources right now to fixing it".  If you have a better suggestion of text we can use to encourage people making feature requests to read the guidelines and understand the level of support we're able to provide, let me know!

**Comment by stubailo** [comment]
I've updated the text to say: "Thanks for the feature request! We welcome discussions about how to make Meteor better. If you haven't yet, check out our wiki page about [feature requests on GitHub](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#feature-requests)." Does that sound good to people? @raix @rclai

**Comment by raix** [comment]
@stubailo I ment adding visuals like italic and the seperation line makes it more clear that its a pasted message - some of us are getting old and senile - well, could be just me - @glasser I'm just glad you guys have the time for feature requests at all, I think its important for the community.

If it was me, ideally I would comment and set the label and have the bot add the std. text accordingly to label. *(without knowing how many std. texts you have, not sure if possible at all)*

-

Btw. I started the https://github.com/MeteorCommunity/discussions/issues to keep track of feature requests, we should prop. migrate FR from there at some point - having it for tech community discussions.

**Comment by mitar** [comment]
What about using a label for this? "Feature request" can be one label, but then "Yes, this would be cool" and "Great, we are going to allocate limited labor resources right now to fixing it" labels would also be cool. Because to me, when you make that reference to the feature request guidelines, I always wonder what exactly are you trying to say. That I missed something in the feature request? That you are confirming it? That you will start working on it? Or that community should start doing pull requests.

For example, it is unclear what you think about this idea:
* is this something MDG thinks could be done, and will do
* is this something MDG thinks could be done, but community should make a pull request, and it will be merged ASAP
* is this something MDG thinks could be done, community should make a pull request, but it will be reviewed and merged when there will be time (pull requests reviews take resources as well)
* do not even bother, MDG does not think this is a good idea for core, make a package/fork
* do not even bother, MDG does not even think this is a good idea for the package

**Comment by stubailo** [comment]
@mitar figuring out which of the five types of feature requests you listed apply to this one is in itself a lot of work.

**Comment by mitar** [comment]
:-)

**Comment by rclai** [comment]
Okay anyways, so looks like another [reactjs-related package](https://atmospherejs.com/avital/react-in-blaze) reached the atmosphere by one of MDG guys @avital. Are you guys going to do some kind of official integration package of sorts?

**Comment by stubailo** [comment]
Right now Avi and Greenspan are building an app with react to test it out, I suspect that package is part of their effort.

**Comment by Goyapa** [comment]
@raix i agree to what you write abbout react-native.

This are some resources i think are very helpful:

[React: CSS in JS – NationJS]
(http://blog.vjeux.com/2014/javascript/react-css-in-js-nationjs.html)

[A Deep Dive into React Native]
(https://www.youtube.com/watch?v=7rDsRXj9-cU)

[Pros and Cons of Facebook's React vs. Web Components (Polymer)]
(https://programmers.stackexchange.com/questions/225400/pros-and-cons-of-facebooks-react-vs-web-components-polymer/237762#237762?newreg=36470115bcbb4d0f9f46d1d5c2751b64)

[Pete Hunt talks Facebook React](https://www.meteor.com/blog/2014/02/14/pete-hunt-talks-facebook-react)

Do you have other resources that could be helpful?

[Ben Newman](http://gotocon.com/chicago-2015/speaker/Ben+Newman) an active React contributor, represents Meteor in the TC39 JavaScript standards process, who initialized [ReactMeteor](https://github.com/reactjs/react-meteor).

**Comment by Rebolon** [comment]
One question about react integration with meteor : does iron-router compatible with react ? should we use react router ? how to manage subscriptions ? A lot of questions that are actually solved by Blaze, but what about React ?

**Comment by rclai** [comment]
If I'm not mistaken, the app that Avi and Greenspan are working on is [here](https://github.com/meteor/rectangles)

**Comment by hellogerard** [comment]
I will add unnecessarily to this thread to say the React is clearly HOT right now. 

Here are all the meteor / react integrations that I'm aware of:

1. https://github.com/reactjs/react-meteor
1. https://github.com/hipertracker/meteor-reactjs/
1. https://github.com/avital/react-in-blaze
1. https://github.com/ccorcos/meteor-reactor
1. https://github.com/mystor/meteor-routecore

**Comment by grigio** [comment]
I add also [Flow Compoments](https://github.com/meteorhacks/flow-components/tree/docs) which is a React -inspired Blaze component lib by @arunoda .

Personally I like Blaze, but React/ReactNative is very hot and a bigger community.

**Comment by arunoda** [comment]
I think this discussion should be not to switch allow to use react. I think there is a good progress on that.
That will help a lot of react developers to choose Meteor over something like gulp+nodejs stack.

I also hope it won't be much harder to mix both Blaze and React. Then that'd be super awesome. That's why we build flow at the first place.

**Comment by MaazAli** [comment]
This would be awesome. Having official react support would take some load off of MDG.

**Comment by Goyapa** [comment]
This is a good blogpost from @tmeasday about React
Creating maintainable interfaces with React & Meteor
http://blog.percolatestudio.com/engineering/reactive-user-interfaces/

**Comment by hitchcott** [comment]
Wanted to +1 this.

Check out my comments on https://github.com/reactjs/react-meteor/issues/41

**Comment by hitchcott** [comment]
Looks like we already have a POC of SSR! :fireworks: 
https://github.com/percolatestudio/react-leaderboard. Great work @zol and @tmeasday!

**Comment by grigio** [comment]
I like Blaze but it isn't much used outside meteor itself, React has a bigger community and ReactNative

**Comment by hellogerard** [comment]
Agree with @grigio. From a coding perspective, I prefer the Blaze way. However, it's starting to get tough to compete with React. Blaze on the server is on the roadmap. Blaze Native is a glimmer at this point (https://groups.google.com/forum/m/#!topic/meteor-core/hcFHexSg1fg)

**Comment by zol** [comment]
@hitchcott don't mind that you deployed react-leaderboard. Thanks for getting involved in the discussion!

**Comment by bobiblazeski** [comment]
1+ for https://github.com/tmeasday/react-leaderboard only react package from the mentioned that has a example  that works  outside the box

**Comment by awatson1978** [comment]
Can anybody explain to me how React isn't simply flavor-of-the-month?  Seems like we had the same discussions over Famo.us and Polymer just a few months ago.  

If anything, it seems like a careful compare/contrast between the three approaches is needed.  What are the common approaches between Famo.us, Polymer, and React?  How do they differ in terms of encapsulation, inheritance, composability, standards compliance, native widget support, render pipeline, use of precompilers, etc?

**Comment by zol** [comment]
@awatson1978 we did our best to explain [here](http://blog.percolatestudio.com/engineering/reactive-user-interfaces/). The fact that facebook, instragram, yahoo, kahn academy and others having been using react in production for 1+ years suggests it may have more substance than just being the flavor of the month.

**Comment by mitar** [comment]
Not necessary. Blaze is not really provided stand-alone. If you have your own existing codebase and you want a fancy reactive UI, react is the way to go. But for Meteor this might not be necessary so. Because react has to work with all other systems, they have to provide that top-level state changing and then propagating changes to children, and then diffing the whole thing. Because Meteor depends on Tracker, this can be minimized. But this also means that others without Meteor have harder time using it.

I like how Blaze/Meteor has separation of concerns, templates and logic. I think that with Blaze 2 discussion there are some good ideas how to have components. And server side rendering is just engineering effort, not that Blaze would not be able to support it.

I think that Blaze already internally use diffing, it applies only changes to the real DOM. This is even better if you have 3rd party code modifying the DOM. Not sure how React reacts to that.

**Comment by awatson1978** [comment]
@zol... and you also gave the same writeup treatment for Famo.us [here](http://blog.percolatestudio.com/engineering/the-future-of-javascript-animation-with-famous/) and were sort of advocating Famo.us component integration just a few months ago.  It just seems like it's just coverage of the latest technology trends to keep the spotlight on Percolate (well done coverage, I might add), rather than necessarily advocating a coherent architecture.  

I'm looking for a discussion like this one, but focused on Meteor:
http://programmers.stackexchange.com/questions/225400/pros-and-cons-of-facebooks-react-vs-web-components-polymer

So what if facebook, instragram, yahoo, kahn academy etc have been using React for a year?  There are other sites that have been using Ember for years.  Or Angular for years.  Or Bootstrap.  Or a thousand other libraries.  

Why should I go through and rewrite dozens of applications with this one particular library?  Native widget support might be the win.  But component reusability?  The ``render()`` approach of React might be performant, but it's code looks like a mess and would be a giant headache to port existing code to.  Why go through all that trouble and still use the HTML precompiler to compile to DOM, and not just built up the DOM one's self?  If I'm going to do that, I might as well use Famo.us.  

Don't get me wrong.  I'm all for React integration, and think it would be an excellent technology to add to the Meteor stack.  But I want to see comparative/quantitative analysis and refactor paths before considering replacing a core library like Blaze for something else.  I've been burned by flavor-of-the-month too many times before in the past few years.

**Comment by murillo128** [comment]
My 2 cents,

I kind of agree with Mitar and in fact we have gone the other way round, 
choosing Blaze standalone (i.e. without Meteor server side) for 
templating and data binding. It is a pity that the standalone Blaze 
library is not activelly supported, but it was not that difficult to 
extract the code and create an npm package 
(https://github.com/eface2face/meteor-client if any one is interested).

I didn't want to use something that required a high learning curve, and 
liked the fact that Blaze/Tracker did exactly what the meant to be doing 
(reactivity, data binding + templating) and nothing else. That allowed 
us to reuse most of the code we had been already using without any 
issue. In fact we even created a jquery plugin( 
https://github.com/eface2face/jquery-meteor-blaze) to make it trivial to 
use:

<script type="text/spacebars"  name="list">
         Below is the listof  items:
         <ol>
         {{#each posts}}
                 <li><input type="checkbox">{{ title }}</li>
         {{/each}}
         </ol>
</script>
<script>
     var  templates=  $("script[type='text/spacebars']").spacebars();
     $("#list")
         .blaze(templates['list'])
         .render({"posts":  [{title:  "first"},{title:  "second"}]);
</script>


In fact, we are now going one step further and started creating our 
self-made old-school web components based on jquery widgets 
(https://github.com/eface2face/jquery-widget-compiler).

My point is that I really think that Blaze and Tracker, due to the fact 
that it has clean interfaces/dependencies has huge possibilities, not 
only on the meteor community, but as standalone library. So, I would 
love to see any real live metrics/comparations between React on 
Blaze/Tracker, to be able to understand what are the caveats/bottlenecks 
existing in the current code and see if we can improve them.

Best regards
Sergio

**Comment by hellogerard** [comment]
Still surprised that https://github.com/mystor/meteor-routecore is not mentioned more. It has a fully functioning, Meteor-React integration TodoMVC example with server-side rendering. 

Not to say that `reactjs:react` won't eventually take it over, especially with MDG working on it. It has a different design, wrapping React's `render` call in an autorun (I actually like the mixin approach better.)

@awatson1978 I take your point about flavor of the month. I understand Polymer much better than I do React right now, so I'm hoping to be able to make an informed comparison over time.

**Comment by mitar** [comment]
Is there any information about interoperability between React and Blaze? Can be React added to otherwise Blaze content and seen as foreign content by Blaze? Can React work with parts of it touched by something else? Or by just having a partial control over DOM?

**Comment by mitar** [comment]
@hellogerard: What do you think about Polymer and its integration with Meteor now?

**Comment by idris** [comment]
I've been [writing an app](https://github.com/asm-products/coinvest-io) with Meteor + React, and I must say the combination is _incredible_!  The pub/sub and `Meteor.call` functionality of Meteor are very similar to the Flux concepts in React, and translate very well.  I'd love to see Meteor embrace React to create a truly magical stack.  Apps will become so much more simple to write.

Oh, and once React Native is out, this is a no-brainer.  Awesome JS stack _and_ a native app? :boom:

**Comment by MaazAli** [comment]
@idris Did you have any noticeable problems? Do you think it can scale well for larger applications, something like [phanime](https://github.com/phanime/phanime)?

**Comment by idris** [comment]
@MaazAli no noticeable problems.  React scales _really well_ for large applications.  It's the brainchild of Facebook, which is a pretty large app :)  It really helps make code more understandable and maintainable, which is especially important if your team becomes more than just a couple of devs.

If you are going to try it, or want to learn more about Meteor+React, I'd suggest signing up for the [Reactiflux Slack](http://reactiflux.com/) and joining the #meteor channel.

**Comment by andrewreedy** [comment]
:+1: for [embracing](https://www.pandastrike.com/posts/20150311-react-bad-idea) the open standards. If an app is shitty, it's going to be equally as shitty in polymer, react, blaze, famous, ember, jquery ui, angular, etc.. If we are moving in a direction it should be less influenced by the current and more influenced by the lighthouse .. (the standards).

**Comment by idris** [comment]
I guess Meteor is relatively template-agnostic right now, since meteor-react works great.  So I guess it makes sense to discuss React Native here.  That would be a relatively big change for Meteor, since it would mean "learn once, write anywhere" instead of "write one, run anywhere".  I think at the very least, having React Native as an option is important.

**Comment by awatson1978** [comment]
Thank you @andrewreedy for posting that excellent, excellent article!  Agreed that this seems more like a forum discussion than a feature request.

**Comment by jonjamz** [comment]
Blaze is becoming cool, but it should be optional. I think it's important to support major existing UI frameworks in a project as large-scoped as Meteor, even while developing your own. A brief comparison between Blaze and the others would help people choose the right one based on their goals.

Overall, I have to agree with @awatson1978 and @andrewreedy here. @idris can you please provide some evidence, even anecdotal, to show how React scales better for larger applications than Blaze or Angular or anything else? And I'm not just talking about performance, here--I'm talking about evolving a high-quality codebase that's easily maintained.

My gut reaction to JSX has been "why the hell would anybody do this?!" Long-established anti-patterns in web development include writing inline HTML in JS and putting JS inside HTML element attributes. Sure, it can be great for a quick demo. It might even perform well. But it's easy to see it evolving into some horrific mass of disorganized, unmaintainable code.

React smells more like a way to attract trendy developers to Facebook than a smart take on best-practices. And certainly, rewriting an existing application to use it would be a bad idea unless there's a serious win somewhere.

**Comment by MaazAli** [comment]
> Overall, I have to agree with @awatson1978 and @andrewreedy here. @idris can you please provide some evidence, even anecdotal, to show how React scales better for larger applications than Blaze or Angular or anything else? And I'm not just talking about performance, here--I'm talking about evolving a high-quality codebase that's easily maintained.

It's more about Blaze vs Ember, React and other front end frameworks. Blaze is super easy to use but I can't see it scale well until we have self-contained components, inheritance and a good way to manage explicit dependencies like import and export (this is more of a meteor issue in general). In my opinion, having globals doesn't scale well in general. Ember by default supports ES6 import and exports among as does React. All of the current ES6 transpilers packaged for meteor cannot properly support imports and exports because of meteor's default way of loading all the files. 

Then you've also got the initial page load and the size of .js file the client receives, as far as I am aware Blaze's generated .js file is fairly large comparatively which again won't scale until we have some type of fast-boot type app rehydration (Ember Fast-boot).

My main concern with Blaze and I guess Meteor in general is being able to export and import modules/components, having explicit dependency management is so much more powerful and clear for other developers than having globals floating all over the place, without this, how do you see meteor scaling well.

**Comment by andrewreedy** [comment]
:+1: for ES6 modules.. (different issue.. new feature request?)

**Comment by jonjamz** [comment]
@MaazAli Blaze already has self-contained component capabilities, although that is relatively new and hasn't been marketed very well. I believe your ES6 module concerns relate to Meteor's bundling/compiling and are not Blaze-specific. Happy to be corrected, though.

You can create nicely maintainable application structures in Meteor today. You just have to put some thought into it and write good JavaScript. You don't need ES6 modules. Of course, support for transpiling those would be nice.

**Comment by mitar** [comment]
> Blaze is becoming cool, but it should seem more optional. I think it's important to support major existing UI frameworks in a project as large-scoped as Meteor even while developing your own. A brief comparison between Blaze and the others would help people choose the right one based on their goals.

The issue is that this is more the question of the ecosystem. Even if Blaze is optional (and it already is), if the library I want to use offers Blaze components, I will have to use them. Or reimplement it.

So instead of trying to make guess which rendering system to use, we should rather figure out a way for them to interoperable. So that I can use both of them in the same app. So how to have a rendering which allows that. I saw examples where it was easy to embed React inside Blaze. But how easy it is to embed Blaze inside of React? If we can abstract this patterns out, then we could make this work.

At the end, all of them are compiled down to the same language: DOM. So we might be able to figure out something here.

Same as it mostly does not matter if my library is written in CoffeeScript or JavaScript for somebody using it from outside.

**Comment by ccorcos** [comment]
I've been [playing around with React recently](https://github.com/ccorcos/meteor-react-utils) and I really like it.

React is actually quite simple. Its just a rendering engine, and its actually very easy to integrate with meteor using a React mixin. I've [created a package for playing around with React that has all sorts of nifty tools](https://github.com/ccorcos/meteor-react-utils) that integrate meteor seamlessly with React. Its not a serious package yet, but you may find it interesting, and maybe you can help me with it.

I can see form Meteor's perspective why they want to keep Blaze. Its super simple and elegant, and for simple things, its amazing. However, React is far more powerful at abstracting user interfaces. So much so, that React Native is practically no different from React for the web! The fact that React does not rely on a markup language gives you so much more freedom in how you design and program your application.

**Comment by raix** [comment]
http://facebook.github.io/react-native/

**Comment by ccorcos** [comment]
!!!!!!!!

**Comment by rclai** [comment]
Oh snap.

**Comment by hitchcott** [comment]
:metal:

**Comment by emmanuelbuah** [comment]
@raix This is awesome!!!

**Comment by stubailo** [comment]
Can't wait for someone to use React Native with Meteor!

**Comment by arunoda** [comment]
With Meteor?
I hope it's for iOS.
You mean with DDP?

**Comment by rclai** [comment]
Right now, it looks like they only have iOS support. Android will come later.

**Comment by emmanuelbuah** [comment]
Looking at the polyfills, localstorage isn't included. I don't know if minimongo uses html5 localstorage or not. If it doesn't then it might be possible to use the default javascript ddp client. If it does, then react native might need to polyfill localstorage for it to work. 

I could be wrong.

**Comment by rclai** [comment]
Wow, if you look at the Github page and you refresh constantly, the stars keep going up. It is just blowing up.

**Comment by ccorcos** [comment]
@emmanuelbuah they have their own ["asyncStorage" API](http://facebook.github.io/react-native/docs/asyncstorage.html#content)

**Comment by raix** [comment]
@stubailo it would be nice if we could add cli commands + build tools via packages :)
That said I'm considering doing a meteor test release of the thing, if anybody up to it we could group to test it out.

**Comment by ccorcos** [comment]
@raix, I'm interested. How are you planning on divvying up the client javascript to React Native?

**Comment by andrewreedy** [comment]
@raix +1 for adding CLI commands via packages!!! You should do a new feature request of that one.

**Comment by dcsan** [comment]
i'm at the F8 conference today and just got a long demo of react native. It looks amazing, the UIs smoothness is just what I want for mobile. if anyone has questions let me know to ask the dev team here.

I haven't used react much myself so this maybe old news but:

For react native there isn't a css-like markup language per se, so all the UI styling is done in JS down to setting font sizes. but it's just JS so you can have your own classes for styling and modify them, eg increase font size for certain devices (no real media queries). so this part would diverge from the desktop version of an app. They have implemented flexbox subset for UI layouts.

They don't support local storage because it is synchronous API.

UI explorer app shows demos of a lot of widgets
https://github.com/facebook/react-native/tree/master/Examples/UIExplorer

Many node libraries are bundled up, they're using NPM as a package manager that compiles the code down to pass to be loaded.

The demo app i saw doesn't have a "reactive data source" in the way that meteor works, but they mostly just use setState. 

Some native API polyfills you can register a callback [when state changes eg example](https://github.com/facebook/react-native/blob/b1827af7dc6a03d687c3ab91bf0c6bc9f868e34f/Examples/UIExplorer/GeolocationExample.js#L55-L57)

```
    this.watchID = navigator.geolocation.watchPosition((lastPosition) => {
      this.setState({lastPosition});
    });
```

**Comment by dcsan** [comment]
they also talked about flow which seems mostly a Typescript like type-checking pre-compile step. In the FB internal build of atom (nuclide) the flow errors come up inline in the editor, right in place. very convenient.

Also announced today but not open yet is GraphQL:
https://facebook.github.io/react/blog/2015/02/20/introducing-relay-and-graphql.html

there isn't really a good server side pairing for React yet. I hope meteor can work towards really tight integration. It still sucks to play with clunky cordova browser apps after seeing the fluidity of the react UX.

**Comment by awatson1978** [comment]
So, just a quick question for those who are diving deeper into React Native... is it built on top of the iOS7 Native Javascript Bridge that was released at the end of 2013?  From everything I'm seeing - the timing, the code samples, who's releasing it, etc - that's what I'm gathering is the case.  But I can't find any explicit documentation to that effect.

**Comment by dcsan** [comment]
react native is built on top of the latest bridge, yes. 
they use JS core on ios and v8 on android.
jscore/ios still doesn't allow JIT code to run, but apple apparently recently updated their developer TOS to allow people to ship JS code to be run later.

["Relay" system](https://facebook.github.io/react/blog/2015/02/20/introducing-relay-and-graphql.html) is their not yet open system that will wrap React components and provide an interface for keeping data up to date, managing subscriptions etc. Perhaps meteor client side could be a drop in for this layer?

graphQL provides a lot of layers of abstraction for managing your data store and queries, and seems a bit daunting for a startup to implement. kind of reminds me of SOAP or something but maybe once you have all those schema type mappings in place it enables you to iterate quicker and safer.

Android version they do have a demo app, but aren't going to open source it until they've released something of their own that is built with it, like the current Ads manager or FB groups apps that are react native.

**Comment by andrewreedy** [comment]
Honestly, after watching the video on Relay and GraphQL .. it sounds like that IS the data transaction layer designed for React .. not to say that Meteor can't be another one. If Meteor decided to go with React as their component system .. I think they will just be chasing Facebook around and end up not having a unique place in the market other than easy app distribution and package management. It sounds like React native is using NPM anyway and I'm sure they are going to come out with the Facebook version of Galaxy for React...  

I think the question here should be how can we take the cool concepts that do make sense for Meteor's unique ecosystem and apply them to make Meteor better rather than chase the herd. The resource agnostic data transport layer ... honestly sounds pretty great! Rhetorical question: How easy is it to set up a new resource with DDP with out using a mongo DSL in the process? .. I think answering that question will be a huge step forward for Meteor.. Also, How easy is it to maintain a large application with Mongo style query subscriptions in the client mixed with Blaze templates? 

Again.. lets try to think about how we can design a better system without throwing in the cards when something new comes along.

**Comment by awatson1978** [comment]
Thanks for confirming that, @dcsan.  If that's the case, my gut instinct is that this FR would be better renamed as 'Add Native Javascript Bridge Support' and added to the Trello board.

**Comment by dcsan** [comment]
React native is the huge difference here. All the other discussion really
doesn't matter if your final mobile app feels like jank.
I would prefer to keep using meteor but if blaze native isn't near term
then the writing is on the wall for meteor. Although I agree strategically
it's hard to see exactly how MDG can insert themselves  into the fb
ecosystem, there isn't really a server side story for react atm. Relay
reminds me of SOAP with all its layers of abstraction.

**Comment by emmanuelbuah** [comment]
+1 to @andrewreedy and @awatson1978. I don't think MDG is in this situation alone. Most of the js platforms/framework are trying to figure out what the benefit are and what to do. A good example is ember cherry picking aspects of react to integrate  into their next release. 

It would be great to hear from members of the MDG team and their general thoughts.

If there is one thing I agree with, its that the current hybride technologies come "close" to native but that it, just close, and the difference between close and native is still huge.

**Comment by dcsan** [comment]
I guess this is turning more into a discussion, so opened a thread on the forums
https://forums.meteor.com/t/react-native-and-meteor/2122

**Comment by emmanuelbuah** [comment]
Spent some time digesting react, react-native and I personally don't think we need to switch to react. There is a lot going on in the react world and no one knows where its going to land. Best we integrate the awesome aspects of react into meteor. Below is my wishlist (not all of them are react related)  

Blaze component
Blaze dom management using virtual dom (performance)
Blaze component event messaging channel - facilitate communication to and from component
Blaze serverside
Meteor Native (similar to react native, built on iOS Native Javascript Bridge)

**Comment by dcsan** [comment]
basically duplicate all the work react team already has done? that's quite a shopping list! 
Meteor native (or rather Blaze native) for ios and android alone is a huge task.

I'd suggest that incorporating the react work would be better, given FB has a huge+expert team and made a huge FOSS contribution already.

FB's "relay" seem to be overblown and not startup friendly, and a perfect fit to be replaced with meteor. If we could instead use react native with the rest of meteors development elegance, that would save many dev years and finally enable competitive mobile apps.

**Comment by emmanuelbuah** [comment]
@dcsan  I understand where you are coming from but unless MDG wants to be chasing facebook around   it doesn't make sense. Think about it, the principles of fb and MDG isn't 1:1. What if facebook comes up with a new toy tomorrow and decide not to support react. What do we do. This is not just taking up react but about the long term out look of MDG. 

To be fair, I meant implementing the suggested features at meteor by leveraging fb's effort just like what Ember is doing. Ember is doing something similar so they don't loose their aim and tightly couple themselves to react. 

Also, react native is specifically for native? MDG is already putting in effort to make cordova it's mobile strategy. My suggest was to shift that strategy towards a meteor-native strategy similar to fb's.  

But in the end, it doesn't matter what we think. MDG will chart its own course. We can only offer our 2 cents. lol

**Comment by ccorcos** [comment]
@dcsan I found that [React works really nicely with Meteor's back end](https://github.com/ccorcos/meteor-react-utils). Relay looks like a pain...

**Comment by MaazAli** [comment]
There seems to be an official package made by react guys (I think he is now part of MDG) [here](https://github.com/reactjs/react-meteor). Supports using react with blaze.

**Comment by ccorcos** [comment]
I don't think Meteor should ditch Blaze. But I think its important to separate technologies a bit more so Meteor can work nicely with different frameworks. 

1. I should be able to use Meteor entirely without Blaze. Then I could use Angular or React if I wanted without having to battle the framework

2. I should be able to use tracker / minimongo / livequery in a client-only script via npm (I'm not sure what else is necessary...) so we can simply require those packages from within React Native if you want to use a Meteor backend with React Native.

3. I should be able use all of the client-side frameworks, including blaze, to create a client-only app.

**Comment by msavin** [comment]
I really like Blaze - its simple and easy to grasp. In some ways, it makes HTML feel even more natural. I think its great for beginners who are starting with Meteor, but official React support would be exciting too.

**Comment by ccorcos** [comment]
I agree with @msavin. Blaze is so awesome for beginners. But I think React does push the limits a little further. I think its best if Meteor were just a little less opinionated about these sorts of things...

**Comment by andrewreedy** [comment]
First +1 to @ccorcos's last comment on options.. (please don't make us use anything **like mongo**)

Sooo I'll throw out some hypotheticals .. What if Facebook (given their man power) makes Relay & Graph QL really easy to use and SUPER efficient and they pair it with some epic node back end architecture ... because.. why not? Then, they throw in an awesome CLI that uploads to their React/Relay app cloud infrastructure .. because thats really not that hard... What then?

The issue here is "Switching to React" .. and my question is: What does that say about Meteor as a platform if we are considering "switching" out core pieces for more built out alternatives from companies like Facebook. I think there needs to be stronger voices in the community and from MDG saying **this** is what we're going to do to make **our** ecosystem just as strong and possibly **easier** to use and closer to the standards of where the web is going... (Like Ember is doing)..

Otherwise people will start to lose faith in the platform .. and this becomes another socketstream ... another derby..

**Comment by msavin** [comment]
It's far from another Derby or SocketStream... Meteor is good as a complete solution and Blaze is really nice. You can get started without knowing anything traditional front-end engineering. I also think MDG's vision is to create a visual builder for the UI's, so integrating with these crazy frameworks may not be their goal.

I'd say the right solution is to develop adapters for alternative clients, such as iOS, Apple WatchOS, Ember, Angular, React, IOT things, etc.

**Comment by andrewreedy** [comment]
Derby (full stack realtime framework).. SocketStream (Node.js based realtime framework) .. Anyway, when you are serious about building scaleable production applications .. "nice" isn't what you are looking for. We need a more well defined components system etc (Ref. @emmanuelbuah's comment for more ideas). 

"You can get started without knowing anything traditional front-end engineering." .. You can get started with anything without knowing how to do it. Front-end engineering is actually really easy to pick up. 

These aren't "crazy" frameworks.. Ember / Facebook's ecosystem are super optimized well designed frameworks which Meteor should be and are in fact concerned with.. given a member of MDG core wrote `reactjs/react-meteor`..

There **are** already solutions to using most of these clients with Meteor (the one's which make sense). The issue here is replacing Blaze with React...

**Comment by jonjamz** [comment]
Have to strongly agree with @andrewreedy here. The answer for Meteor is *certainly not* to differentiate itself out of the league of serious production-grade frameworks (and into the league of Squarespace, or whatever you're describing @msavin).

**Comment by msavin** [comment]
@jonjamz maybe Andrew can join us for that coffee and we'll discuss it

(Spoiler: I think Meteor can and probably/hopefully will be visualized..)

**Comment by zimt28** [comment]
I also think that "switching" is wrong. But it would be cool if we could choose, what we want to use. Be it Blaze (default), React, Riot, Angular ..

**Comment by ccorcos** [comment]
Has anyone made any attempts to use React Native with Meteor?

**Comment by raix** [comment]
@ccorcos I've been reading source code in the weekend, theres 3 angles of attack I think:
* Add react native as a first class package with cli integration - eg. add `$ meteor add-platform react-native-ios` and `$ meteor run ios-react` + server-side rendering etc.
* Integrate Blaze with the react-native-bridge
* Create a Blaze-native - where we replace the react parts in the react-native and get the bridge running via the meteor cli

So I think it would be nice if we could give them all a go and evaluate on the results - theres pro/cons to it all - personally I'm 100% for a Blaze-native but it takes some ressources - a pure react solution would prop. be easiest.

I'm in NYC 18-19/4 if you guys want to talk meteor/react-native in person

**Comment by jonjamz** [comment]
@raix I would be happy to meet up in NYC and talk about Meteor in person. Anybody else? Perhaps @msavin @awatson1978 would be available?

**Comment by LuisHerranz** [comment]
I am with @raix here. I think BlazeNative should be the way to go.

And doing BlazeNative it's not starting from scratch like somebody said here.
It's porting ReactNative to be used with Blaze syntax and with Meteor in general. Facebook already did much of the work and open sourced so other projects can use it.

I really think MDG should jump in and let us know what they think.

**Comment by jonjamz** [comment]
@ccorcos @raix I'd be happy to contribute to a Blaze Native project if MDG isn't already working on something similar. Will see if anybody chimes in.

**Comment by stubailo** [comment]
@jonjamz MDG here, we are not currently working on Blaze Native, it would be interesting to see what people here could come up with!

**Comment by ccorcos** [comment]
@raix I'm more than happy to help. (1) seems like a good first step but I don't have a clear vision of how we'd do that. My thinking was along the lines of creating npm packages that import meteor into a react-native project. I think that would serve a dual purpose of integrating with React Native, but also making meteor more friendly with the outside community. Although I could see from the MDG perspective, they'd probably want to keep everything in meteor and galaxy-compatible...

Let me know when you start on this and I'm happy to help.

**Comment by stubailo** [comment]
Speaking as a non-MDG representative here - I think if I were to go about this project I would start by making a React Native frontend that can connect over DDP to a Meteor server backend. That way you can use all of the developer tools Facebook has released for React Native, and just import the frontend parts of Meteor that you need, for example accounts, MiniMongo, DDP client, etc.

**Comment by martijnwalraven** [comment]
I also think connecting React Native to a Meteor server over DDP is probably the way to go for now. I’ve been working on an iOS DDP client (https://github.com/martijnwalraven/meteor-ios), for which I reimplemented parts of the Meteor logic in native code. I'm not sure yet how much of that would be useful when integrating with React Native, as it probably makes more sense to run Meteor packages like minimongo in a JavaScriptCore context instead. But reusing parts of the native DDP protocol code might make sense.

The programming model could be similar to the recently released Parse integration and the proposed observe() API for React:
https://github.com/ParsePlatform/ParseReact
https://github.com/facebook/react/issues/3398

I'd be happy to help, but how much time I’ll get to spend on this depends on whether I’ll be able to find clients willing to sponsor parts of the work.

**Comment by emmanuelbuah** [comment]
I agree with @stubailo and @martijnwalraven. I have always thought a lot of these discussion will be mute if the client-side aspect of meteor (minimongo/storage, ddp) was solid/official for ios and android, just like the web. @martijnwalraven I like how parse introduced their data layer through React mixin. 

I think building our famous meteor todo app in React-native incorporating the needed meteor client-side (MiniMongo, DDP client, accounts) but from my understanding (base on comments from others), these piece aren't standalone. So the question is, how do we get standalone packages of these clients-side meteor libs to get started.

I more that happy to help as im currently working on huqoo's ios app.

**Comment by raix** [comment]
@jonjamz let's mail regarding time/place in NYC would be great to have more in the minimeetup, please feel free to join in,

I've created a simple Trello board for tracking things and added a few of you - please reach out if interested. https://trello.com/b/V73P9Z5t/meteor-react

@stubailo I've added your suggestion - would be great to cover some ground, it could actually be a first step in/before a Meteor react release (I find it hard to live without the packages system managing versions...)

@ccorcos I've tested react-native just a bit in the weekend + read half through the code - had some outage due to the DDOS Attack on github.

**Comment by stubailo** [comment]
@raix I'm not exactly sure what you mean by the following:

> I've added your suggestion - would be great to cover some ground, it could actually be a first step in/before a Meteor react release (I find it hard to live without the packages system managing versions...)

But it sounds important - could you elaborate?

**Comment by raix** [comment]
@stubailo The Meteor into React-native packager could help React developers get started with Meteor. At the same time we could figure out the best way of having the Meteor packages work with React-native - info that would make it trivial to create a meteor release based on React-native, allowing users to take advantage of isopack/build etc.

**Comment by raix** [comment]
We could potentially have spacebars/templating even blaze work react-native

**Comment by trusktr** [comment]
@MaazAli 

> All of the current ES6 transpilers packaged for meteor cannot properly support imports and exports because of meteor's default way of loading all the files.

But you can use a `prepublish` script in your package.json to convert it all to CommonJS modules for publishing on npmjs.org, then it'll work just fine.

**Comment by trusktr** [comment]
> Add react native as a first class package with cli integration - eg. add $ meteor add-platform react-native-ios and $ meteor run ios-react + server-side rendering etc.

Would server-side rendering even be useful with react-native?

**Comment by dcsan** [comment]
what is the first stage to get react native hooked up to meteor data sources? I would prefer to start with that. I like Blaze but react API is also fine. But I would prefer to take the shortest route to a native renderer and meteor for server and maybe client side collections, models, mini-mongo.

Blaze native, would mean re-implementing all of the code facebook has to support react view layer, seems like a whole separate task. Only MDG or someone sufficiently funded to do just that has the resources. It could also be a second step once the basic react-native=meteor data is working.

@raix happy to help if i can, pls add me to the trello board. pending other projects but i would love to see this working.

**Comment by emmanuelbuah** [comment]
Guys, I thought this - http://www.appcelerator.com/blog/2015/02/react-native-reaction/ was an interesting read from the guys at appcelerator. It's interesting to note that http://www.appcelerator.com/platform/apps/ is is pretty much the same as react-native (philosophy wise) with "react" being the main diff. Where appcelerator differs is its focus on single component api for both iOS & Android as much as possible. 
Surprised I didn't know about http://www.appcelerator.com/platform/apps/. This validates the use case of the ability to use meteor clients-side (ddp, accounts, minimongo, etc) in other front-end platforms. Thus, we can use it to build appcelerator apps as well as react-native apps.

**Comment by dcsan** [comment]
last time i tried appcelerator it was pretty janky, esp on android. 
they announced a "hyperloop" native layer a while back aka TI.next
not sure if that's what their 3.5 is based on, but it's been a long time coming
http://www.appcelerator.com/blog/2014/12/introducing-hal/
https://github.com/appcelerator/hyperloop/commits/master

**Comment by LuisHerranz** [comment]
From my understanding, the two great/new things about ReactNative are:

 - Javascript runs in another thread, so the UI is always fluid.
 - They implemented flexbox, so you can layout like you do in HTML.

Does anybody have any experience with NativeScript?
https://www.nativescript.org/
Their approach at wrapping the native APIs looks great.

**Comment by emmanuelbuah** [comment]
@LuisHerranz Thanks for sharing nativescript. Haven't heard of it. It looks pretty solid and again, similar if not the same as react-native - markup + js marshaled to native call and visa versa.

I going to spend sometime this weekend on an ios native meteor todo app on react-native and hopefully, nativescript.

**Comment by LuisHerranz** [comment]
Another way to go can be Asteroid:
http://mondora.github.io/asteroid/

If we can get it working inside ReactNative (or NativeScript) we could access collections, reactive variables, accounts, server methods... with ddp.

What do you think?

**Comment by emmanuelbuah** [comment]
Great option since we don't have an official readily available meteor client-side lib. @raix Suggest at https://trello.com/c/OFeD6Lrd/4-react-ddp-package we use the meteor generated clientside bundle from an empty project but asteroid works.

**Comment by hellogerard** [comment]
I put together some React examples for a talk I did at the local Cincy meetup. One of them is React + Asteroid. Here's the repo:

https://github.com/Differential/todos-playground

My intent is to experiment with react-native, but haven't had the time yet.

**Comment by ccorcos** [comment]
Asteroid doesn't use Tracker which is pretty lame...

**Comment by ccorcos** [comment]
1. It occurred to me that we'd lose all the benefits of hot code pushes if we go the route of an nom package.

2. [Meteor-ios](https://github.com/martijnwalraven/meteor-ios) is actually very impressive. Perhaps using this in combination with React Native would be a good option.

**Comment by riaan53** [comment]
Hi All,

I have posted something on the meteor forums that might be of interest. https://forums.meteor.com/t/virtual-dom-a-virtual-dom-and-diffing-algorithm/2388

Regards,
Riaan

**Comment by petehunt** [comment]
Wanted to chime in here -- someone has started working on websocket support for react-native here: https://github.com/SanderSpies/react-native-1/tree/WebSocket

I believe once websocket support in react-native lands, starting to use ddp from react-native would be pretty trivial as long as the ddp client doesn't have a dependency on the browser DOM. If the ddp client can get by with only XMLHTTPRequest then it should work today.

Getting Blaze to work in react-native would be a challenge. You can either look at the react-native bridge protocol (which is platform-agnostic) or do something like this on top of vanilla React http://wix.github.io/react-templates/

**Comment by mitar** [comment]
I couldn't wait, so I just did it. Just released: https://github.com/peerlibrary/meteor-blaze-components

Blaze Components for Meteor are a system for easily developing complex UI elements that need to be reused around your Meteor app.

Feedback welcome. :-)

**Comment by ccorcos** [comment]
nice job @mitar!

**Comment by mitar** [comment]
See live tutorial: http://components.meteor.com/

**Comment by rclai** [comment]
Is there a JavaScript version? Lol.

**Comment by dcsan** [comment]
wow, that's a huge contribution, and nicely written tutorial too.
now we just need the native implementation!

**Comment by arunoda** [comment]
@mitar nicely done. Problems oriented tutorial is nice.

you may need to deploy the app with production mode: http://components.meteor.com/
It's takes a lot of times to load.

**Comment by grigio** [comment]
A problem with React is the npm distribution of the [3rd part components/deps](https://github.com/callemall/material-ui/issues/331), but it already work fine for custom nested components. Here is an example of a Meteor React reusable component https://github.com/grigio/meteor-react-overlay

**Comment by spyhole** [comment]
Wanted to +1 for this and ReactNative!

**Comment by babakness** [comment]
ReactNative is great, it makes data consumption a breeze. 

NativeScript is worth consideration as well. The main attraction with NativeScript is that it can use native APIs without wrappers. 

See NativeScript.org

**Comment by hharnisc** [comment]
Hey everyone, hacked this together today 

https://github.com/hharnisc/react-native-meteor

Uses `Objective-DDP` to talk to Meteor and then sends events to JavaScript using [React eventDispatcher](http://facebook.github.io/react-native/docs/nativemodulesios.html#sending-events-to-javascript). This solution seems to work pretty well, and could be a viable solution for production code if some of the generic bits were abstracted. 

I still think what @petehunt suggested about waiting for websocket Polyfills is a better solution -- [status on the work here](https://github.com/facebook/react-native/issues/619).  Once there are websockets a whole bunch of NPM packages become available (https://www.npmjs.com/package/ddp, https://www.npmjs.com/package/asteroid) to name a few.

**Comment by hharnisc** [comment]
I've hacked together another example using a [WebSocket polyfill](https://github.com/facebook/react-native/pull/890) and a customized version of the [DDP](https://www.npmjs.com/package/ddp) NPM module.

https://github.com/hharnisc/react-native-meteor-websocket-polyfill

The advantage to this approach is you can keep your head in the JavaScript and don't need to install any third party Objective-C pods

Here is the connection/subscription/observe flow from the **previous example**

```javascript
DDPClient.connectWithURL('ws://localhost:3000/websocket');
DDPClient.subscribe('publicLists');

var currentRows = Immutable.List.of();
var that = this;

var addedObserver = RCTDeviceEventEmitter.addListener('added', function(data) {
  currentRows = currentRows.push(data);
  that.updateRows(currentRows.toArray());
});

var changedObserver = RCTDeviceEventEmitter.addListener('changed', function(data) {
  currentRows = currentRows.map(function(row){
    if (row._id === data._id) {
      return data
    } else {
      return row
    }
  });
  that.updateRows(currentRows.toArray());
});

var removedObserver = RCTDeviceEventEmitter.addListener('removed', function(data) {
  currentRows = currentRows.filter(function(row) {
    return row._id !== data._id
  })
  that.updateRows(currentRows.toArray());
});
```

Here's what it looks like now

```javascript
var ddpClient = new DDPClient({url: 'ws://localhost:3000/websocket'});

ddpClient.connect(() => ddpClient.subscribe('publicLists'));

// observe the lists collection
var observer = ddpClient.observe("lists");
observer.added = () => this.updateRows(_.cloneDeep(_.values(ddpClient.collections.lists)));
observer.changed = () => this.updateRows(_.cloneDeep(_.values(ddpClient.collections.lists)));
observer.removed = () => this.updateRows(_.cloneDeep(_.values(ddpClient.collections.lists)));
```

I'm planning on giving a lightning talk on this at the next devshop this month and talk about what was involved in making this work. Getting pretty close to ideal!

**Comment by guilhermedecampo** [comment]
Really cool @hharnisc !

**Comment by Maxim-Filimonov** [comment]
@hharnisc just curious if polyfill works would it be possible to use minimongo and other parts of client side meteor instead of DDPClient? The case i'm thinking of is using meteor client side javascript part as replacement for Relay and GraphQL .

**Comment by hharnisc** [comment]
Hey @Maxim-Filimonov great question! I very briefly tried pulling out the `DDP` and `Minimongo` client side packages from Meteor. There were a couple places in the code with dependencies on the `document` (things like `onready`) that kept me from moving forward.

With some more work it may be possible to use them. It would take either modifying the Meteor package to not depend on the document, or stubbing out a document with React Native. @Slava any thoughts on removing references on those client side packages?

I think this is worth bringing up that the devshop!

**Comment by Maxim-Filimonov** [comment]
@hharnisc Can you point a specific files where it uses document? I did a quick search from current dev meteor and i can't find any onready / document references in DDP or minimongo packages

**Comment by hharnisc** [comment]
If you walk down the dependency tree it eventually depends on Meteor... Which includes the reference (there may be others, but I don't remember exactly which ones)



- Harrison

**Comment by hharnisc** [comment]
PR has for adding the Websocket polyfill has been accepted! https://github.com/facebook/react-native/pull/890

**Comment by zol** [comment]
:+1:

**Comment by hellogerard** [comment]
:boom: :metal:

**Comment by emmanuelbuah** [comment]
react-native socketio come through today - https://github.com/kirkness/react-native-swift-socketio. Thought I share.

**Comment by grigio** [comment]
@Rebolon You can use `flow-router`, `react-router` or `iron-router` and subscribe at route level or react component level.

**Comment by omeid** [comment]
:+1: :100:

**Comment by aleclarson** [comment]
I have converted most of the Meteor packages to be compatible with React Native.

Let me know what you guys think (and if you find any errors). :smile:

-

- [meteor-client@1.1.7](https://github.com/aleclarson/meteor-client) [required]

- [mongo@1.1.1](https://github.com/aleclarson/mongo)

- [tracker@1.0.8](https://github.com/aleclarson/tracker)

- [check@1.0.6](https://github.com/aleclarson/check)

-

- [ddp@1.2.0](https://github.com/aleclarson/ddp)

- [minimongo@1.0.9](https://github.com/aleclarson/minimongo)

- [json@1.0.4](https://github.com/aleclarson/json)

- [ejson@1.0.7](https://github.com/aleclarson/ejson)

- [geojson@1.0.4](https://github.com/aleclarson/geojson)

- [random@1.0.4](https://github.com/aleclarson/random)

- [mongo-id@1.0.0](https://github.com/aleclarson/mongo-id)

- [reload@1.1.4](https://github.com/aleclarson/reload)

- [retry@1.0.4](https://github.com/aleclarson/retry)

- [diff-sequence@1.0.0](https://github.com/aleclarson/diff-sequence)

- [id-map@1.0.4](https://github.com/aleclarson/id-map)

- [ordered-dict@1.0.4](https://github.com/aleclarson/ordered-dict)

- [underscore@1.5.2](https://github.com/aleclarson/underscore)

**Comment by ccorcos** [comment]
@aleclarson, do you have an example app using these packages?

**Comment by aleclarson** [comment]
@ccorcos Not yet. Not much testing has been done either. So far, I've successfully connected React Native's `WebSocket` polyfill to a default Meteor server (via the new `Meteor.connect('ws://localhost:3000/websocket')`). I'm optimistic that everything should work as expected. I'll be testing more today and get back to you! :+1:

**Comment by ccorcos** [comment]
Awesome! Any thoughts on integrating with the Meteor build tool like cordova?

**Comment by aleclarson** [comment]
@ccorcos Ideally, React Native's packager would be a plugin for Isobuild. But for now, I decided that converting the client-side Meteor code to be compatible with React Native's packager would be less work.

**Comment by grigio** [comment]
Another alternative way to embrace React without wrapper packages is to embrace the React + ES2015 +Webpack ecosystem https://github.com/jedwards1211/meteor-webpack-react

**Comment by rclai** [comment]
@grigio is it possible to achieve React Native with Meteor by using Webpack or is there any massaging of Meteor codebase that needs to happen?

**Comment by nicholasalanbrown** [comment]
Differential just wrote a great blog post about combining Meteor and React Native: http://blog.differential.com/meteor-react-native-learning-from-experience/ 

I'm personally hoping official support for React Native comes soon...

**Comment by aleclarson** [comment]
@rclai After taking a look at [jedwards1211/meteor-webpack-react](https://github.com/jedwards1211/meteor-webpack-react), it seems as though it only supports React (not React Native). Until official support comes (if it ever does), I think [aleclarson/meteor-client](https://github.com/aleclarson/meteor-client) is the only way to use Meteor in React Native. I would love to be proved wrong, though.

**Comment by RonaldV** [comment]
As other people have suggested if you want to keep Blaze, it might be interesting to integrate with NativeScript. In that case it probably would be best to integrate with Blaze 2 (having components and data/state passing?). I my opinion having a native integration point is pretty important. Cordova just doesn't cut it. You'll need to have a separation of the render layer from the application layer in Blaze so it can integrate with NativeScript.

**Comment by jolic** [comment]
+1 for NativeScript integration

**Comment by emmanuelbuah** [comment]
I totally agree. Nativescript is in a much better place than React native
or any of the other js to native toolset. I have been using it for a while
and its very simple. Still working on meteor integration.

**Comment by jolic** [comment]
@emmanuelbuah 
Do you have a working example on github?

**Comment by emmanuelbuah** [comment]
@jolic I have a working sample using rest (not on github yet) but not a ddp sample. I'm currently working on a ddp client for Nativescript - https://github.com/emmanuelbuah/nativescript.ddp. I will welcome any help I can get. I can add you to the gitter thread as well.

**Comment by ccorcos** [comment]
Looks like React Native might be coming soon :)

https://forums.meteor.com/t/webpack-compiler-inside-meteor-es6-modules-hot-reload-and-code-splitting

**Comment by tonyxiao** [comment]
@ccorcos Can you explain how does webpack compiler help with bringing ReactNative to Meteor?

**Comment by ccorcos** [comment]
It just gives you better control over your files and where they get compiled -- and React Native uses webpack as a build system.

**Comment by tonyxiao** [comment]
So you can package all meteor dependencies via webpack into a single bundle and then use it within ReactNative?

**Comment by ccorcos** [comment]
Yeah -- I'm not sure how the meteor dependencies would work... You'd have to use NPM.
